## Week 4, Lab 1: Predicting Left-Handedness from Psychological Factors
> Author: Matt Brems

We can sketch out the data science process as follows:
1. Define the problem.
2. Obtain the data.
3. Explore the data.
4. Model the data.
5. Evaluate the model.
6. Answer the problem.

We'll walk through a full data science problem in this lab. 

---
## Step 1: Define The Problem.

You're currently a data scientist working at a university. A professor of psychology is attempting to study the relationship between personalities and left-handedness. They have tasked you with gathering evidence so that they may publish.

Specifically, the professor says "I need to prove that left-handedness is caused by some personality trait. Go find that personality trait and the data to back it up."

As a data scientist, you know that any real data science problem must be **specific** and **conclusively answerable**. For example:
- Bad data science problem: "What is the link between obesity and blood pressure?"
    - This is vague and is not conclusively answerable. That is, two people might look at the conclusion and one may say "Sure, the problem has been answered!" and the other may say "The problem has not yet been answered."
- Good data science problem: "Does an association exist between obesity and blood pressure?"
    - This is more specific and is conclusively answerable. The problem specifically is asking for a "Yes" or "No" answer. Based on that, two independent people should both be able to say either "Yes, the problem has been answered" or "No, the problem has not yet been answered."
- Excellent data science problem: "As obesity increases, how does blood pressure change?"
    - This is very specific and is conclusively answerable. The problem specifically seeks to understand the effect of one variable on the other.

### 1. In the context of the left-handedness and personality example, what are three specific and conclusively answerable problems that you could answer using data science? 

> You might find it helpful to check out the codebook in the repo for some inspiration.

Answer: Does an association exist between personality and left-handedness?

---
## Step 2: Obtain the data.

### 2. Read in the file titled "data.csv."
> Hint: Despite being saved as a .csv file, you won't be able to simply `pd.read_csv()` this data!

In [2]:
import numpy as np
import pandas as pd

df = pd.read_csv('data.csv', delimiter='\t')
df

Q1  Q2  Q3  Q4  Q5  Q6  Q7  Q8  Q9  Q10  ...  country  fromgoogle  \
0      4   1   5   1   5   1   5   1   4    1  ...       US           2   
1      1   5   1   4   2   5   5   4   1    5  ...       CA           2   
2      1   2   1   1   5   4   3   2   1    4  ...       NL           2   
3      1   4   1   5   1   4   5   4   3    5  ...       US           2   
4      5   1   5   1   5   1   5   1   3    1  ...       US           2   
...   ..  ..  ..  ..  ..  ..  ..  ..  ..  ...  ...      ...         ...   
4179   3   5   4   5   2   4   2   2   2    5  ...       US           1   
4180   1   5   1   5   1   4   2   4   1    4  ...       US           1   
4181   3   2   2   4   5   4   5   2   2    5  ...       PL           2   
4182   1   3   4   5   1   3   3   1   1    3  ...       US           2   
4183   2   5   3   3   5   3   4   3   1    5  ...       NZ           2   

      engnat  age  education  gender  orientation  race  religion  hand  
0          1   22          3       1            1     3         2     3  
1          1   14          1       2            2     6         1     1  
2          2   30          4       1            1     1         1     2  
3          1   18          2       2            5     3         2     2  
4          1   22          3       1            1     3         2     3  
...      ...  ...        ...     ...          ...   ...       ...   ...  
4179       1   18          2       1            1     6         2     1  
4180       1   18          2       2            1     3         2     1  
4181       2   22          2       1            1     6         1     1  
4182       1   16          1       2            5     1         1     1  
4183       1   22          3       2            4     6         1     1  

[4184 rows x 56 columns]

### 3. Suppose that, instead of us giving you this data in a file, you were actually conducting a survey to gather this data yourself. From an ethics/privacy point of view, what are three things you might consider when attempting to gather this data?
> When working with sensitive data like sexual orientation or gender identity, we need to consider how this data could be used if it fell into the wrong hands!

Answer: Indicate clearly that there is no good or bad personality

---
## Step 3: Explore the data.

### 4. Conduct exploratory data analysis on this dataset.
> If you haven't already, be sure to check out the codebook in the repo, as that will help in your EDA process.

In [3]:
df.isnull().sum()
df.dtypes
df.describe()

Q1           Q2           Q3           Q4           Q5  \
count  4184.000000  4184.000000  4184.000000  4184.000000  4184.000000   
mean      1.962715     3.829589     2.846558     3.186902     2.865440   
std       1.360291     1.551683     1.664804     1.476879     1.545798   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       1.000000     3.000000     1.000000     2.000000     1.000000   
50%       1.000000     5.000000     3.000000     3.000000     3.000000   
75%       3.000000     5.000000     5.000000     5.000000     4.000000   
max       5.000000     5.000000     5.000000     5.000000     5.000000   

                Q6           Q7           Q8           Q9          Q10  ...  \
count  4184.000000  4184.000000  4184.000000  4184.000000  4184.000000  ...   
mean      3.672084     3.216539     3.184512     2.761233     3.522945  ...   
std       1.342238     1.490733     1.387382     1.511805     1.242890  ...   
min       0.000000     0.000000     0.000000     0.000000     0.000000  ...   
25%       3.000000     2.000000     2.000000     1.000000     3.000000  ...   
50%       4.000000     3.000000     3.000000     3.000000     4.000000  ...   
75%       5.000000     5.000000     4.000000     4.000000     5.000000  ...   
max       5.000000     5.000000     5.000000     5.000000     5.000000  ...   

          testelapse   fromgoogle       engnat           age    education  \
count    4184.000000  4184.000000  4184.000000   4184.000000  4184.000000   
mean      479.994503     1.576243     1.239962     30.370698     2.317878   
std      3142.178542     0.494212     0.440882    367.201726     0.874264   
min         7.000000     1.000000     0.000000     13.000000     0.000000   
25%       186.000000     1.000000     1.000000     18.000000     2.000000   
50%       242.000000     2.000000     1.000000     21.000000     2.000000   
75%       324.250000     2.000000     1.000000     27.000000     3.000000   
max    119834.000000     2.000000     2.000000  23763.000000     4.000000   

            gender  orientation         race     religion         hand  
count  4184.000000  4184.000000  4184.000000  4184.000000  4184.000000  
mean      1.654398     1.833413     5.013623     2.394359     1.190966  
std       0.640915     1.303454     1.970996     2.184164     0.495357  
min       0.000000     0.000000     0.000000     0.000000     0.000000  
25%       1.000000     1.000000     5.000000     1.000000     1.000000  
50%       2.000000     1.000000     6.000000     2.000000     1.000000  
75%       2.000000     2.000000     6.000000     2.000000     1.000000  
max       3.000000     5.000000     7.000000     7.000000     3.000000  

[8 rows x 55 columns]

---
## Step 4: Model the data.

### 5. Suppose I wanted to use Q1 - Q44 to predict whether or not the person is left-handed. Would this be a classification or regression problem? Why?

Answer: classification because the y variable is categorical.

### 6. We want to use $k$-nearest neighbors to predict whether or not a person is left-handed based on their responses to Q1 - Q44. Before doing that, however, you remember that it is often a good idea to standardize your variables. In general, why would we standardize our variables? Give an example of when we would standardize our variables.

Answer: To correct data which are not normal

### 7. Give an example of when we might not standardize our variables.

Answer: It will affect the predictability of the data since natrually collected data is not standardized.

### 8. Based on your answers to 6 and 7, do you think we should standardize our predictor variables in this case? Why or why not?

Answer: I think since this is a survey data, it might be better not to standardize it.

### 9. We want to use $k$-nearest neighbors to predict whether or not a person is left-handed. What munging/cleaning do we need to do to our $y$ variable in order to explicitly answer this question? Do it.

Answer: 

In [4]:
#Need to treat those people who are both-handed as left-handed
df.loc[df.hand == 0, 'left_handed'] = 0
df.loc[df.hand == 1, 'left_handed'] = 0
df.loc[df.hand == 2, 'left_handed'] = 1
df.loc[df.hand == 3, 'left_handed'] = 1
df.left_handed = df.left_handed.astype(int)

In [5]:
pd.crosstab(df.hand, df.left_handed)

left_handed     0    1
hand                  
0              11    0
1            3542    0
2               0  452
3               0  179

### 10. The professor for whom you work suggests that you set $k = 4$. In this specific case, why might this be a bad idea?

Answer: Odd number would be better so to avoid having a tie situation.

### 11. Let's *(finally)* use $k$-nearest neighbors to predict whether or not a person is left-handed!

> Be sure to create a train/test split with your data!

> Create four separate models, one with $k = 3$, one with $k = 5$, one with $k = 15$, and one with $k = 25$.

> Instantiate and fit your models.

In [7]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.preprocessing import StandardScaler
from sklearn.datasets import load_iris

X = df.iloc[:,0:44]
y = df.left_handed

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=.2, random_state=42, stratify=y)

knn3 = KNeighborsClassifier(n_neighbors=3)
knn5 = KNeighborsClassifier(n_neighbors=5)
knn15 = KNeighborsClassifier(n_neighbors=15)
knn25 = KNeighborsClassifier(n_neighbors=25)

knn3.fit(X_train, y_train)
knn3train = knn3.score(X_train, y_train)
knn3test = knn3.score(X_test, y_test)

knn5.fit(X_train, y_train)
knn5train = knn5.score(X_train, y_train)
knn5test = knn5.score(X_test, y_test)

knn15.fit(X_train, y_train)
knn15train = knn15.score(X_train, y_train)
knn15test = knn15.score(X_test, y_test)

knn25.fit(X_train, y_train)
knn25train = knn25.score(X_train, y_train)
knn25test = knn25.score(X_test, y_test)


print('KNN03 ' + 'train: ' + str(knn3train) + ' test: ' + str(knn3test))
print('KNN05 ' + 'train: ' + str(knn5train) + ' test: ' + str(knn5test))
print('KNN15 ' + 'train: ' + str(knn15train) + ' test: ' + str(knn15test))
print('KNN25 ' + 'train: ' + str(knn25train) + ' test: ' + str(knn25test))

KNN03 train: 0.870331640274873 test: 0.8088410991636799
KNN05 train: 0.8518075888855692 test: 0.8291517323775388
KNN15 train: 0.8497161637287123 test: 0.8482676224611708
KNN25 train: 0.8491186136838961 test: 0.8494623655913979


Being good data scientists, we know that we might not run just one type of model. We might run many different models and see which is best.

### 12. We want to use logistic regression to predict whether or not a person is left-handed. Before we do that, let's check the [documentation for logistic regression in sklearn](http://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegression.html). Is there default regularization? If so, what is it? If not, how do you know?

Answer: Yes, it is applied by default

### 13. We want to use logistic regression to predict whether or not a person is left-handed. Before we do that, should we standardize our features?

Answer:

### 14. Let's use logistic regression to predict whether or not the person is left-handed.


> Be sure to use the same train/test split with your data as with your $k$-NN model above!

> Create four separate models, one with LASSO and $\alpha = 1$, one with LASSO and $\alpha = 10$, one with Ridge and $\alpha = 1$, and one with Ridge and $\alpha = 10$. *(Hint: Be careful with how you specify $\alpha$ in your model!)*

> Instantiate and fit your models.

In [12]:
from sklearn.linear_model import Lasso, LassoCV, Ridge
from sklearn.linear_model import LogisticRegression

model1 = LogisticRegression(penalty='l1', C=1.0, solver='liblinear', class_weight='balanced')
model1.fit(X_train, y_train)
lasso1train = model1.score(X_train, y_train)
lasso1test = model1.score(X_test, y_test)

model2 = LogisticRegression(penalty='l1', C=10.0, solver='liblinear', class_weight='balanced')
model2.fit(X_train, y_train)
lasso10train = model2.score(X_train, y_train)
lasso10test = model2.score(X_test, y_test)

model3 = LogisticRegression(penalty='l2', C=1.0, solver='liblinear', class_weight='balanced')
model3.fit(X_train, y_train)
ridge1train = model3.score(X_train, y_train)
ridge1test = model3.score(X_test, y_test)

model4 = LogisticRegression(penalty='l2', C=10.0, solver='liblinear', class_weight='balanced')
model4.fit(X_train, y_train)
ridge10train = model4.score(X_train, y_train)
ridge10test = model4.score(X_test, y_test)

print('LASSO = 1 ' + 'train: ' + str(lasso1train) + ' test: ' + str(lasso1test))
print('LASSO = 10 ' + 'train: ' + str(lasso10train) + ' test: ' + str(lasso10test))
print('RIDGE = 1 ' + 'train: ' + str(ridge1train) + ' test: ' + str(ridge1test))
print('RIDGE = 10 ' + 'train: ' + str(ridge10train) + ' test: ' + str(ridge10test))

LASSO = 1 train: 0.5873916940543771 test: 0.5651135005973715
LASSO = 10 train: 0.5876904690767851 test: 0.5651135005973715
RIDGE = 1 train: 0.5873916940543771 test: 0.5639187574671446
RIDGE = 10 train: 0.5873916940543771 test: 0.5651135005973715


---
## Step 5: Evaluate the model(s).

### 15. Before calculating any score on your data, take a step back. Think about your $X$ variable and your $Y$ variable. Do you think your $X$ variables will do a good job of predicting your $Y$ variable? Why or why not? What impact do you think this will have on your scores?

I am skeptical that they are good predictors unless there are prior research done that have established that personality has impact on Y.

### 16. Using accuracy as your metric, evaluate all eight of your models on both the training and testing sets. Put your scores below. (If you want to be fancy and generate a table in Markdown, there's a [Markdown table generator site linked here](https://www.tablesgenerator.com/markdown_tables#).)
- Note: Your answers here might look a little weird. You didn't do anything wrong; that's to be expected!

KNN03 train: 0.870331640274873 test: 0.8088410991636799<br>
KNN05 train: 0.8518075888855692 test: 0.8291517323775388<br>
KNN15 train: 0.8497161637287123 test: 0.8482676224611708<br>
KNN25 train: 0.8491186136838961 test: 0.8494623655913979<br>
LASSO = 1 train: 0.5873916940543771 test: 0.5651135005973715<br>
LASSO = 10 train: 0.5876904690767851 test: 0.5651135005973715<br>
RIDGE = 1 train: 0.5873916940543771 test: 0.5639187574671446<br>
RIDGE = 10 train: 0.5873916940543771 test: 0.5651135005973715<br>

### 17. In which of your $k$-NN models is there evidence of overfitting? How do you know?

It seems that is some overfitting for KNN03 and KNN05. The train figure is better than the test figure.

### 18. Broadly speaking, how does the value of $k$ in $k$-NN affect the bias-variance tradeoff? (i.e. As $k$ increases, how are bias and variance affected?)

It seems that with higher k, there is less bias and more variance.

### 19. If you have a $k$-NN model that has evidence of overfitting, what are three things you might try to do to combat overfitting?

I could increase the number of k in the model, perform standardization to the data or find more data.

### 20. In which of your logistic regression models is there evidence of overfitting? How do you know?

They all seems to have some form of overfitting. The train figure is higher than the test figure.

### 21. Broadly speaking, how does the value of $C$ in logistic regression affect the bias-variance tradeoff? (i.e. As $C$ increases, how are bias and variance affected?)

Increase C is likely to increase bias and reduce variance.

### 22. For your logistic regression models, play around with the regularization hyperparameter, $C$. As you vary $C$, what happens to the fit and coefficients in the model? What do you think this means in the context of this specific problem?

I don't observe any much difference in adjusting the hyperparameter C.

### 23. If you have a logistic regression model that has evidence of overfitting, what are three things you might try to do to combat overfitting?

Simplify the model (i.e. reduce some of the variables), use regularisation or use more training data.

---
## Step 6: Answer the problem.

### 24. Suppose you want to understand which psychological features are most important in determining left-handedness. Would you rather use $k$-NN or logistic regression? Why?

I will use logistic regression then I can use the coefficient to determine which psychological feature is more important.

### 25. Select your logistic regression model that utilized LASSO regularization with $\alpha = 1$. Interpret the coefficient for `Q1`.

LASSO = 1 train: 0.5873916940543771 test: 0.5651135005973715

### 26. If you have to select one model overall to be your *best* model, which model would you select? Why?
- Usually in the "real world," you'll fit many types of models but ultimately need to pick only one! (For example, a client may not understand what it means to have multiple models, or if you're using an algorithm to make a decision, it's probably pretty challenging to use two or more algorithms simultaneously.) It's not always an easy choice, but you'll have to make it soon enough. Pick a model and defend why you picked this model!

I would use logistic regression because it seems to provide a more realistic figure on how the X variables are really contributing to the effect on the Y variable.

### 27. Circle back to the three specific and conclusively answerable questions you came up with in Q1. Answer one of these for the professor based on the model you selected!

In [53]:
importance = model.coef_
featurecoeff = []
indexlist = []
n=1

for i,j in enumerate(importance):
    for k in j:
        featurecoeff.append(k)
        indexlist.append(n)
        n += 1

result = pd.DataFrame(list(zip(indexlist, featurecoeff)),
              columns=['Feature','Coefficient'])

result["AbsCoeff"] = result.Coefficient.abs()

result.sort_values(by="AbsCoeff",ascending=False)

Feature  Coefficient  AbsCoeff
42       43    -0.225646  0.225646
26       27     0.201441  0.201441
34       35     0.133089  0.133089
22       23    -0.127432  0.127432
38       39    -0.098989  0.098989
37       38     0.082631  0.082631
24       25     0.079527  0.079527
7         8    -0.076769  0.076769
21       22    -0.073275  0.073275
2         3     0.069689  0.069689
25       26     0.069071  0.069071
39       40    -0.051287  0.051287
20       21    -0.049107  0.049107
15       16     0.045636  0.045636
16       17     0.045247  0.045247
8         9    -0.043071  0.043071
40       41    -0.037284  0.037284
32       33     0.036941  0.036941
29       30    -0.034726  0.034726
5         6    -0.033422  0.033422
4         5     0.033240  0.033240
19       20    -0.033130  0.033130
43       44     0.033075  0.033075
3         4     0.031444  0.031444
30       31    -0.031303  0.031303
28       29     0.028648  0.028648
0         1     0.028344  0.028344
35       36    -0.027661  0.027661
1         2    -0.022788  0.022788
6         7     0.022439  0.022439
12       13    -0.022086  0.022086
41       42    -0.019428  0.019428
14       15    -0.016946  0.016946
13       14    -0.016153  0.016153
27       28     0.014890  0.014890
10       11     0.014540  0.014540
36       37     0.013663  0.013663
11       12     0.009527  0.009527
17       18    -0.009325  0.009325
23       24     0.008060  0.008060
18       19    -0.006344  0.006344
33       34    -0.005717  0.005717
9        10    -0.003900  0.003900
31       32     0.001431  0.001431

The top 5 personality questions that are more importance to determin left-handedness are<br>
Q42	I bake sweets just for myself sometimes.<br>
Q26	I jump up and down in excitement sometimes.<br>
Q34	I have kept a personal journal.<br>
Q22	I save the letters I get.<br>
Q38	I leave nice notes for people now and then.<br>

### BONUS:
Looking for more to do? Probably not - you're busy! But if you want to, consider exploring the following. (They could make for a blog post!)
- Create a visual plot comparing training and test metrics for various values of $k$ and various regularization schemes in logistic regression.
- Rather than just evaluating models based on accuracy, consider using sensitivity, specificity, etc.
- In the context of predicting left-handedness, why are unbalanced classes concerning? If you were to re-do this process given those concerns, what changes might you make?
- Fit and evaluate a generalized linear model other than logistic regression (e.g. Poisson regression).
- Suppose this data were in a `SQL` database named `data` and a table named `inventory`. What `SQL` query would return the count of people who were right-handed, left-handed, both, or missing with their class labels of 1, 2, 3, and 0, respectively? (You can assume you've already logged into the database.)